proposals/lip-0052.md

```diff
@@ -362,712 +379,1832 @@ The module provides the following commands to modify the NFT store.
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_UNLOCK,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics = [module, nftID]
+    )
+
+    def emitFailedUnlockEvent(
+        module: Module,
+        nftID: NFTID,
+        result: uint32
+    ) -> None:
+
+        emitPersistentEvent(
+            module = MODULE_NAME_NFT,
+            name = EVENT_NAME_UNLOCK,
+            data = {
+                "module": module,
+                "nftID": nftID,
+                "result": result
+            },
+            topics = [module, nftID]
+        )
 ```
 
-#### transfer
+#### setAttributes
 
-This function transfers ownership of NFTs within one chain.
+This function modifies the attributes of an NFT.
 
 ```python
-transfer(senderAddress, recipientAddress, nftID):
-    nftID = canonicalNFTID(nftID)
-    if (NFTStore(nftID).lockingModuleID != NFT_NOT_LOCKED
-        or NFTStore(nftID).owner != senderAddress):   
-        transfer fails
+def setAttributes(
+    module: Module,
+    nftID: NFTID,
+    attributes: Attributes
+) -> None:
+
+    if NFTStore(nftID) does not exist:
+        emitFailedSetAttributesEvent(module, nftID, attributes, RESULT_NFT_DOES_NOT_EXIST)
+        raise Exception("NFT substore entry does not exist.")
+
+    if getChainID(nftID) != OWN_CHAIN_ID:
+        emit FailedSetAttributesEvent(module, nftID, attributes, RESULT_NFT_NOT_NATIVE)
+        raise Exception("NFT is not native to the chain.")
+
+    attributes(nftID, module) = attributes
+
+    emitEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_SET_ATTRIBUTES,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "attributes": attributes,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics = [module, nftID]
+    )
+
+    def emitFailedSetAttributesEvent(
+        module: Module,
+        nftID: NFTID,
+        attributes: Attributes,
+        result: uint32
+    ) -> None:
+
+        emitPersistentEvent(
+            module = MODULE_NAME_NFT,
+            name = EVENT_NAME_SET_ATTRIBUTES,
+            data = {
+                "module": module,
+                "nftID": nftID,
+                "attributes": attributes,
+                "result": result
+            },
+            topics = [module, nftID]
+        )
+```
 
-    deleteUserEntry(senderAddress, nftID)
-    createUserEntry(recipientAddress, nftID)
-    NFTStore(nftID).owner = recipientAddress
+#### transfer
+
+This function transfers ownership of an NFT within one chain.
+
+```python
+def transfer(
+    senderAddress: Address,
+    recipientAddress: Address,
+    nftID: NFTID
+) -> None:
+
+    if NFTStore(nftID) does not exist:
+        emitFailedTransferEvent(senderAddress, recipientAddress, nftID, RESULT_NFT_DOES_NOT_EXIST)
+        raise Exception("NFT store entry does not exist.")
+
+    if getNFTOwner(nftID) == OWN_CHAIN_ID:
```

Comment:
Same as above, this check is superfluous but we could keep it for extra error granularity, in which case we should fix the check `len(getNFTOwner(nftID)) == LENGTH_CHAIN_ID`